crates/runtime/src/instance/allocator/pooling/uffd.rs

```diff
@@ -0,0 +1,833 @@
+                                        memory_index,
+                                        page_index,
+                                    )?;
+                                }
+                                None => {
+                                    log::trace!("out of bounds memory access at {:p}", access_addr);
+
+                                    // Record the guard page fault with the instance so it can be reset later.
+                                    instance.record_guard_page_fault(
+                                        page_addr,
+                                        len,
+                                        reset_guard_page,
+                                    );
+                                    wake_guard_page_access(&uffd, page_addr, len)?;
+                                }
+                            }
+                        }
+                        Some(AddressLocation::StackPage {
+                            page_addr,
+                            len,
+                            index,
+                            guard_page,
+                        }) => {
+                            log::trace!(
+                                "handling fault in stack {} at address {:p}",
+                                index,
+                                access_addr,
+                            );
+
+                            if guard_page {
+                                // Logging as trace as stack guard pages might be a trap condition in the future
+                                log::trace!("stack overflow fault at {:p}", access_addr);
+
+                                // Mark the stack as having a faulted guard page
+                                // The next time the stack is used the guard page will be reset
+                                faulted_stack_guard_pages[index].store(true, Ordering::SeqCst);
+                                wake_guard_page_access(&uffd, page_addr, len)?;
+                                continue;
+                            }
+
+                            // Always zero stack pages
+                            uffd.zeropage(page_addr as _, len, true).map_err(|e| {
+                                format!(
+                                    "failed to zero page at {:p} with length {}: {}",
+                                    page_addr, len, e
+                                )
+                            })?;
+                        }
+                        None => {
+                            return Err(format!(
+                                "failed to locate fault address {:p} in registered memory regions",
+                                access_addr
+                            ));
+                        }
+                    }
+                }
+            }
+            Some(_) => continue,
+            None => break,
+        }
+    }
+
+    log::trace!("fault handler thread has successfully terminated");
```

Comment:
This thread just loops on a blocking read on uffd, which doesn't leave any mechanism to tear it down. With the change I suggested above, this is unreachable.

Reply:
As discussed in Slack, the fault handler will monitor for the regions being unmapped and when the count of regions being monitored hits zero, it should gracefully terminate.  This does require a slightly newer kernel version (4.11) compare to Lucet's uffd implementation, however.

The hope with doing it this way is that we get around the weird "POLLIN but can't read" issue we're seeing from Lucet's non-blocking implementation by having blocking reads 🤞